File: GTMAppAuthSwift/Tests/ObjCIntegration/GTMOAuth2KeychainCompatibilityTests.m
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

@import AppAuthCore;
@import GTMAppAuthSwift;
@import TestHelpers;

@interface GTMOAuth2KeychainCompatibilityTests : XCTestCase

@property (nonatomic) GTMAppAuthFetcherAuthorization *expectedAuthorization;

@end

@implementation GTMOAuth2KeychainCompatibilityTests

- (void)setUp {
  self.expectedAuthorization =
      [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:OIDAuthState.testInstance
                                                serviceProvider:GTMTestingConstants.testServiceProvider
                                                         userID:GTMTestingConstants.userID
                                                      userEmail:GTMTestingConstants.testEmail
                                            userEmailIsVerified:@"y"];
  [super setUp];
}

- (NSString *)expectedPersistenceResponseString {
  return [NSString stringWithFormat:@"access_token=%@&refresh_token=%@&scope=%@&serviceProvider=%@&userEmail=foo%%40foo.com&userEmailIsVerified=y&userID=%@", GTMTestingConstants.testAccessToken, GTMTestingConstants.testRefreshToken, GTMTestingConstants.testScope2, GTMTestingConstants.testServiceProvider, GTMTestingConstants.testUserID];
}

- (void)testPersistenceResponseString {
  NSString *testPersistenceResponseString =
      [GTMOAuth2KeychainCompatibility persistenceResponseStringForAuthState:self.expectedAuthorization];
  XCTAssertNotNil(testPersistenceResponseString);
  XCTAssertEqualObjects([self expectedPersistenceResponseString], testPersistenceResponseString);
}

- (void)testAuthStateForPersistenceString {
  NSError *error;
  GTMOAuth2KeychainCompatibility *oauth2Compat = [[GTMOAuth2KeychainCompatibility alloc] init];
  GTMAppAuthFetcherAuthorization *testPersistAuth =
      [oauth2Compat authStateForPersistenceString:[self expectedPersistenceResponseString]
                                         tokenURL:GTMTestingConstants.testTokenURL
                                      redirectURI:GTMTestingConstants.testRedirectURI
                                         clientID:GTMTestingConstants.testClientID
                                     clientSecret:GTMTestingConstants.testClientSecret
                                            error:&error];
  XCTAssertNil(error);
  XCTAssertEqual(testPersistAuth.authState.scope, self.expectedAuthorization.authState.scope);
  XCTAssertEqualObjects(testPersistAuth.authState.lastTokenResponse.accessToken,
                        self.expectedAuthorization.authState.lastTokenResponse.accessToken);
  XCTAssertEqualObjects(testPersistAuth.authState.refreshToken,
                        self.expectedAuthorization.authState.refreshToken);
  XCTAssertEqual(testPersistAuth.authState.isAuthorized,
                 self.expectedAuthorization.authState.isAuthorized);
  XCTAssertEqualObjects(testPersistAuth.serviceProvider,
                        self.expectedAuthorization.serviceProvider);
  XCTAssertEqualObjects(testPersistAuth.userID, self.expectedAuthorization.userID);
  XCTAssertEqualObjects(testPersistAuth.userEmail, self.expectedAuthorization.userEmail);
  XCTAssertEqual(testPersistAuth.userEmailIsVerified,
                 self.expectedAuthorization.userEmailIsVerified);
  XCTAssertEqual(testPersistAuth.canAuthorize, self.expectedAuthorization.canAuthorize);
}

- (void)testAuthStateForPersistenceStringThrows {
  NSError *error;
  GTMOAuth2KeychainCompatibility *oauth2Compat = [[GTMOAuth2KeychainCompatibility alloc] init];
  GTMAppAuthFetcherAuthorization *testPersistAuth =
      [oauth2Compat authStateForPersistenceString:[self expectedPersistenceResponseString]
                                         tokenURL:GTMTestingConstants.testTokenURL
                                      redirectURI:@""
                                         clientID:GTMTestingConstants.testClientID
                                     clientSecret:GTMTestingConstants.testClientSecret
                                            error:&error];
  XCTAssertNotNil(error);
  XCTAssertEqual(error.code, GTMKeychainStoreErrorCodeFailedToConvertRedirectURItoURL);
}

@end

File: GTMAppAuthSwift/Tests/GTMOAuth2KeychainCompatibilityTests.swift
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import XCTest
import AppAuthCore
@testable import GTMAppAuthSwift

class GTMOAuth2KeychainCompatibilityTests: XCTestCase {
  private lazy var testPersistenceString: String = {
    return "access_token=\(Constants.testAccessToken)&refresh_token=\(Constants.testRefreshToken)&scope=\(Constants.testScope2)&serviceProvider=\(Constants.testServiceProvider)&userEmail=foo%40foo.com&userEmailIsVerified=y&userID=\(Constants.testUserID)"
  }()
  private let keychainHelper = KeychainHelperFake()
  private lazy var keychain: GTMKeychain = {
    return GTMKeychain(keychainHelper: keychainHelper)
  }()
  private var expectedAuthorization: GTMAppAuthFetcherAuthorization {
    GTMAppAuthFetcherAuthorization(
      authState: OIDAuthState.testInstance(),
      serviceProvider: Constants.testServiceProvider,
      userID: Constants.testUserID,
      userEmail: Constants.testEmail,
      userEmailIsVerified: "y"
    )
  }

  override func setUp() {
    super.setUp()
    GTMOAuth2KeychainCompatibility.keychain = keychain
  }

  override func tearDown() {
    super.tearDown()
    keychainHelper.passwordStore.removeAll()
    keychainHelper.useDataProtectionKeychain = false
    GTMOAuth2KeychainCompatibility.keychain = nil
  }

  func testPersistenceResponseString() {
    let response = GTMOAuth2KeychainCompatibility.persistenceResponseStringForAuthorization(
      expectedAuthorization
    )
    guard let response = response else {
      return XCTFail("Response shouldn't be nil")
    }
    XCTAssertEqual(response, testPersistenceString)
  }

  func testSaveOAuth2Authorization() throws {
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedAuthorization,
      forKeychainItemName: Constants.testKeychainItemName
    )
  }

  func testSaveOAuth2AuthorizationThrowsError() {
    let emptyItemName = ""
    XCTAssertThrowsError(
      try GTMOAuth2KeychainCompatibility.save(
        authorization: expectedAuthorization,
        forKeychainItemName: emptyItemName
      )
    ) { thrownError in
      XCTAssertEqual(
        thrownError as? GTMKeychainError,
        GTMKeychainError.noService
      )
    }
  }

  func testRemoveOAuth2Authorization() throws {
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedAuthorization,
      forKeychainItemName: Constants.testKeychainItemName
    )
    try GTMOAuth2KeychainCompatibility.removeAuthorizationFromKeychain(forItemName: Constants.testKeychainItemName)
  }

  func testRemoveOAuth2AuthorizationThrowsError() {
    let unsavedItemName = "unsavedItemName"
    XCTAssertThrowsError(
      try GTMOAuth2KeychainCompatibility.removeAuthorizationFromKeychain(forItemName: unsavedItemName)
    ) { thrownError in
      XCTAssertEqual(
        thrownError as? GTMKeychainError,
        .failedToDeletePasswordBecauseItemNotFound(itemName: unsavedItemName)
      )
    }
  }

  func testAuthorizeFromKeychainForName() throws {
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedAuthorization,
      forKeychainItemName: Constants.testKeychainItemName
    )
    let testAuth = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
      forItemName: Constants.testKeychainItemName,
      tokenURL: Constants.testTokenURL,
      redirectURI: Constants.testRedirectURI,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )

    XCTAssertEqual(testAuth.authState.scope, expectedAuthorization.authState.scope)
    XCTAssertEqual(
      testAuth.authState.lastTokenResponse?.accessToken,
      expectedAuthorization.authState.lastTokenResponse?.accessToken
    )
    XCTAssertEqual(testAuth.authState.refreshToken, expectedAuthorization.authState.refreshToken)
    XCTAssertEqual(testAuth.authState.isAuthorized, expectedAuthorization.authState.isAuthorized)
    XCTAssertEqual(testAuth.serviceProvider, expectedAuthorization.serviceProvider)
    XCTAssertEqual(testAuth.userID, expectedAuthorization.userID)
    XCTAssertEqual(testAuth.userEmail, expectedAuthorization.userEmail)
    XCTAssertEqual(testAuth.userEmailIsVerified, expectedAuthorization.userEmailIsVerified)
    XCTAssertEqual(testAuth.canAuthorize, expectedAuthorization.canAuthorize)
  }

  func testAuthorizeFromKeychainForNameThrowsError() throws {
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedAuthorization,
      forKeychainItemName: Constants.testKeychainItemName
    )
    let badRedirectURI = ""
    XCTAssertThrowsError(
      _ = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
        forItemName: Constants.testKeychainItemName,
        tokenURL: Constants.testTokenURL,
        redirectURI: badRedirectURI,
        clientID: Constants.testClientID,
        clientSecret: Constants.testClientSecret
      )
    ) { thrownError in
      XCTAssertEqual(
        thrownError as? GTMOAuth2KeychainCompatibility.Error,
        .failedToConvertRedirectURItoURL(badRedirectURI)
      )
    }
  }

  func testAuthorizeFromKeychainForPersistenceStringFailedWithBadURI() {
    let badURI = ""

    XCTAssertThrowsError(
      try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
        forPersistenceString: testPersistenceString,
        tokenURL: Constants.testTokenURL,
        redirectURI: badURI,
        clientID: Constants.testClientID,
        clientSecret: Constants.testClientSecret
      )
    ) { thrownError in
      XCTAssertEqual(
        thrownError as? GTMOAuth2KeychainCompatibility.Error,
        .failedToConvertRedirectURItoURL(badURI)
      )
    }
  }

  func testAuthorizeFromKeychainForPersistenceString() throws {
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedAuthorization,
      forKeychainItemName: Constants.testKeychainItemName
    )

    let testPersistAuth = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
      forPersistenceString: testPersistenceString,
      tokenURL: Constants.testTokenURL,
      redirectURI: Constants.testRedirectURI,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )

    XCTAssertEqual(testPersistAuth.authState.scope, expectedAuthorization.authState.scope)
    XCTAssertEqual(
      testPersistAuth.authState.lastTokenResponse?.accessToken,
      expectedAuthorization.authState.lastTokenResponse?.accessToken
    )
    XCTAssertEqual(testPersistAuth.authState.refreshToken, expectedAuthorization.authState.refreshToken)
    XCTAssertEqual(testPersistAuth.authState.isAuthorized, expectedAuthorization.authState.isAuthorized)
    XCTAssertEqual(testPersistAuth.serviceProvider, expectedAuthorization.serviceProvider)
    XCTAssertEqual(testPersistAuth.userID, expectedAuthorization.userID)
    XCTAssertEqual(testPersistAuth.userEmail, expectedAuthorization.userEmail)
    XCTAssertEqual(testPersistAuth.userEmailIsVerified, expectedAuthorization.userEmailIsVerified)
    XCTAssertEqual(testPersistAuth.canAuthorize, expectedAuthorization.canAuthorize)
  }

  func testAuthorizeFromKeychainMatchesForNameAndPersistenceString() throws {
    let expectedPersistAuth = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
      forPersistenceString: testPersistenceString,
      tokenURL: Constants.testTokenURL,
      redirectURI: Constants.testRedirectURI,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedPersistAuth,
      forKeychainItemName: Constants.testKeychainItemName
    )

    let testPersistAuth = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
      forItemName: Constants.testKeychainItemName,
      tokenURL: Constants.testTokenURL,
      redirectURI: Constants.testRedirectURI,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )

    XCTAssertEqual(testPersistAuth.authState.scope, expectedPersistAuth.authState.scope)
    XCTAssertEqual(
      testPersistAuth.authState.lastTokenResponse?.accessToken,
      expectedPersistAuth.authState.lastTokenResponse?.accessToken
    )
    XCTAssertEqual(testPersistAuth.authState.refreshToken, expectedPersistAuth.authState.refreshToken)
    XCTAssertEqual(testPersistAuth.authState.isAuthorized, expectedPersistAuth.authState.isAuthorized)
    XCTAssertEqual(testPersistAuth.serviceProvider, expectedPersistAuth.serviceProvider)
    XCTAssertEqual(testPersistAuth.userID, expectedPersistAuth.userID)
    XCTAssertEqual(testPersistAuth.userEmail, expectedPersistAuth.userEmail)
    XCTAssertEqual(testPersistAuth.userEmailIsVerified, expectedPersistAuth.userEmailIsVerified)
    XCTAssertEqual(testPersistAuth.canAuthorize, expectedPersistAuth.canAuthorize)
  }

  func testAuthorizeFromKeychainUsingGoogleOAuthProviderInformation() throws {
    let expectedPersistAuth = try GTMOAuth2KeychainCompatibility.authorizeFromKeychain(
      forPersistenceString: testPersistenceString,
      tokenURL: Constants.testTokenURL,
      redirectURI: Constants.testRedirectURI,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )
    try GTMOAuth2KeychainCompatibility.save(
      authorization: expectedPersistAuth,
      forKeychainItemName: Constants.testKeychainItemName
    )

    let testAuthorization = try GTMOAuth2KeychainCompatibility.authForGoogleFromKeychain(
      forKeychainItemName: Constants.testKeychainItemName,
      clientID: Constants.testClientID,
      clientSecret: Constants.testClientSecret
    )

    XCTAssertEqual(testAuthorization.authState.scope, expectedPersistAuth.authState.scope)
    XCTAssertEqual(
      testAuthorization.authState.lastTokenResponse?.accessToken,
      expectedPersistAuth.authState.lastTokenResponse?.accessToken
    )
    XCTAssertEqual(testAuthorization.authState.refreshToken, expectedPersistAuth.authState.refreshToken)
    XCTAssertEqual(testAuthorization.authState.isAuthorized, expectedPersistAuth.authState.isAuthorized)
    XCTAssertEqual(testAuthorization.serviceProvider, expectedPersistAuth.serviceProvider)
    XCTAssertEqual(testAuthorization.userID, expectedPersistAuth.userID)
    XCTAssertEqual(testAuthorization.userEmail, expectedPersistAuth.userEmail)
    XCTAssertEqual(testAuthorization.userEmailIsVerified, expectedPersistAuth.userEmailIsVerified)
    XCTAssertEqual(testAuthorization.canAuthorize, expectedPersistAuth.canAuthorize)
  }
}
